learningequality/kolibri
# Assign quizzes to individual learners

### Summary


Implements the ability to assign individual learners per the requirement listed in https://www.notion.so/learningequality/UX-UI-design-499b292122a142a1b23124d38a76c420

One notable difference between the mock-up (https://www.figma.com/proto/TcSczscKh4zLTypFJh2Or4/0.13-Lesson-%26-quiz-updates?node-id=176%3A151&scaling=scale-down) and the implementation is that there is no "User type" filter - every user in this context is a Learner so it doesn't need to be there.

### Reviewer guidance


_Please test this using a previously used KOLIBRI_HOME folder. Copy it somewhere and delete it after testing. This PR requires that a new data model be associated with quizzes. When you create a NEW quiz - the logic handles it there. However, there is additional logic that ensures that previously created quizzes (which were created prior to the logic that now exists to include the association upon quiz creation) will also have the appropriate association created._

__Step 1: Assign an Individual Learner__

- Go to Coach > Class > Plan > Quizzes > Quiz Details for a quiz created before running this PR.
- Ensure that the Quiz is "Open"
- Go to the Options > Edit Details
- Uncheck any previously checked Groups if there are any.
- See that the "Individual learners" checkbox exists.
- Click it and see that the Learners assigned to this class are listed there.
- Click one or more of the checkboxes to select a Learner - remember the username of that Learner.
- While the Individual Learners table is open - click a group and see that the learners in the table are disabled if they are part of the selected group.
- Test the filter input for the table please!
- Save

__Step 2: See that the Learner is Assigned__

- In a separate browser window (Incognito / Private) login as the Learner to whom the quiz was assigned.
- See that the learner may take the quiz.
- See that in the Quiz Status sidebar (In Quiz Details) that the Group and Learners names are listed under "Recipients" - if you have 3 other groups selected, you won't see learners' names but you'll see "Group 1, Group2 and N others" - but if you only have one group and some individual learners selected, it should read like "Group 1, user 1 and N others". If it is ONLY individual learners selected, it should be names of learners.

__Step 3: Try to break assigning learners__

- Return to the browser where you are logged in as the Coach.
- Return to the "Edit Details" page for the same quiz.
- Now, select a Group. If one doesn't exist, go back and create a group and assign some but not all learners to it.
- See that the table of Individual learners is now filtered so that the members of the selected group are not visible.
- Now - select several learners individually or as a part of a group and verify that they are able to see the quiz (or not be able to see it) as expected.
- Do the same for several other quizzes as well.

__Step 4: Create a new quiz__

- Go to Coach > Class > Plan > Quizzes > New Quiz
- Create a new quiz
- Replay "Step 1" and "Step 2" for the newly created quiz. 


__Step 5: Side Effects__

- Create a quiz and assign it to a group - then assign it to a number of individual learners. Remember the number of individual learners.
- Go through everything in the Quiz Reports - take quizzes as Individual Learners and not Invidiual learners - are the number reported properly?
- Ensure that the Class Home is reflective of the IndividualLearners and their behavior as well as that of the learners in normal groups.

__Step 5.5: Side Effects (Again)__

- Do Step 5, but ONLY assign individual learners - no other groups.
- Make sure everything works as expected.

### References


https://www.figma.com/proto/TcSczscKh4zLTypFJh2Or4/0.13-Lesson-%26-quiz-updates?node-id=176%3A151&scaling=scale-down

https://github.com/learningequality/clearinghouse/issues/422

https://github.com/learningequality/clearinghouse/issues/352

----

### Contributor Checklist


PR process:


Testing:


### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md
